File: server/storage.ts
import { 
  Application, 
  InsertApplication, 
  DataRequest, 
  InsertDataRequest, 
  QuestionAnalysis, 
  InsertQuestionAnalysis, 
  ToolConnector, 
  InsertToolConnector, 
  DataCollectionSession, 
  InsertDataCollectionSession, 
  AuditResult, 
  InsertAuditResult 
} from "@shared/schema";

export interface IStorage {
  // Application methods
  createApplication(application: InsertApplication): Promise<Application>;
  getApplicationById(id: number): Promise<Application | null>;
  getAllApplications(): Promise<Application[]>;
  updateApplication(id: number, application: Partial<InsertApplication>): Promise<Application | null>;
  deleteApplication(id: number): Promise<boolean>;

  // Data request methods
  createDataRequest(dataRequest: InsertDataRequest): Promise<DataRequest>;
  getDataRequestById(id: number): Promise<DataRequest | null>;
  getDataRequestsByApplicationId(applicationId: number): Promise<DataRequest[]>;
  updateDataRequest(id: number, dataRequest: Partial<InsertDataRequest>): Promise<DataRequest | null>;
  deleteDataRequest(id: number): Promise<boolean>;

  // Question analysis methods
  createQuestionAnalysis(analysis: InsertQuestionAnalysis): Promise<QuestionAnalysis>;
  getQuestionAnalysisById(id: number): Promise<QuestionAnalysis | null>;
  getQuestionAnalysesByApplicationId(applicationId: number): Promise<QuestionAnalysis[]>;
  updateQuestionAnalysis(id: number, analysis: Partial<InsertQuestionAnalysis>): Promise<QuestionAnalysis | null>;
  deleteQuestionAnalysis(id: number): Promise<boolean>;

  // Tool connector methods
  createToolConnector(connector: InsertToolConnector): Promise<ToolConnector>;
  getToolConnectorById(id: number): Promise<ToolConnector | null>;
  getToolConnectorsByApplicationId(applicationId: number): Promise<ToolConnector[]>;
  updateToolConnector(id: number, connector: Partial<InsertToolConnector>): Promise<ToolConnector | null>;
  deleteToolConnector(id: number): Promise<boolean>;

  // Data collection session methods
  createDataCollectionSession(session: InsertDataCollectionSession): Promise<DataCollectionSession>;
  getDataCollectionSessionById(id: number): Promise<DataCollectionSession | null>;
  getDataCollectionSessionsByApplicationId(applicationId: number): Promise<DataCollectionSession[]>;
  updateDataCollectionSession(id: number, session: Partial<InsertDataCollectionSession>): Promise<DataCollectionSession | null>;
  deleteDataCollectionSession(id: number): Promise<boolean>;

  // Audit result methods
  createAuditResult(result: InsertAuditResult): Promise<AuditResult>;
  getAuditResultById(id: number): Promise<AuditResult | null>;
  getAuditResultsByApplicationId(applicationId: number): Promise<AuditResult[]>;
  updateAuditResult(id: number, result: Partial<InsertAuditResult>): Promise<AuditResult | null>;
  deleteAuditResult(id: number): Promise<boolean>;
}

// Memory storage implementation
class MemStorage implements IStorage {
  private applications: Map<number, Application> = new Map();
  private dataRequests: Map<number, DataRequest> = new Map();
  private questionAnalyses: Map<number, QuestionAnalysis> = new Map();
  private toolConnectors: Map<number, ToolConnector> = new Map();
  private dataCollectionSessions: Map<number, DataCollectionSession> = new Map();
  private auditResults: Map<number, AuditResult> = new Map();
  
  private nextId = 1;

  // Application methods
  async createApplication(application: InsertApplication): Promise<Application> {
    const newApp: Application = {
      id: this.nextId++,
      ...application,
      createdAt: new Date(),
    };
    this.applications.set(newApp.id, newApp);
    return newApp;
  }

  async getApplicationById(id: number): Promise<Application | null> {
    return this.applications.get(id) || null;
  }

  async getAllApplications(): Promise<Application[]> {
    return Array.from(this.applications.values());
  }

  async updateApplication(id: number, application: Partial<InsertApplication>): Promise<Application | null> {
    const existing = this.applications.get(id);
    if (!existing) return null;
    
    const updated = { ...existing, ...application };
    this.applications.set(id, updated);
    return updated;
  }

  async deleteApplication(id: number): Promise<boolean> {
    return this.applications.delete(id);
  }

  // Data request methods
  async createDataRequest(dataRequest: InsertDataRequest): Promise<DataRequest> {
    const newRequest: DataRequest = {
      id: this.nextId++,
      ...dataRequest,
      uploadedAt: new Date(),
    };
    this.dataRequests.set(newRequest.id, newRequest);
    return newRequest;
  }

  async getDataRequestById(id: number): Promise<DataRequest | null> {
    return this.dataRequests.get(id) || null;
  }

  async getDataRequestsByApplicationId(applicationId: number): Promise<DataRequest[]> {
    return Array.from(this.dataRequests.values()).filter(dr => dr.applicationId === applicationId);
  }

  async updateDataRequest(id: number, dataRequest: Partial<InsertDataRequest>): Promise<DataRequest | null> {
    const existing = this.dataRequests.get(id);
    if (!existing) return null;
    
    const updated = { ...existing, ...dataRequest };
    this.dataRequests.set(id, updated);
    return updated;
  }

  async deleteDataRequest(id: number): Promise<boolean> {
    return this.dataRequests.delete(id);
  }

  // Question analysis methods
  async createQuestionAnalysis(analysis: InsertQuestionAnalysis): Promise<QuestionAnalysis> {
    const newAnalysis: QuestionAnalysis = {
      id: this.nextId++,
      ...analysis,
      createdAt: new Date(),
    };
    this.questionAnalyses.set(newAnalysis.id, newAnalysis);
    return newAnalysis;
  }

  async getQuestionAnalysisById(id: number): Promise<QuestionAnalysis | null> {
    return this.questionAnalyses.get(id) || null;
  }

  async getQuestionAnalysesByApplicationId(applicationId: number): Promise<QuestionAnalysis[]> {
    return Array.from(this.questionAnalyses.values()).filter(qa => qa.applicationId === applicationId);
  }

  async updateQuestionAnalysis(id: number, analysis: Partial<InsertQuestionAnalysis>): Promise<QuestionAnalysis | null> {
    const existing = this.questionAnalyses.get(id);
    if (!existing) return null;
    
    const updated = { ...existing, ...analysis };
    this.questionAnalyses.set(id, updated);
    return updated;
  }

  async deleteQuestionAnalysis(id: number): Promise<boolean> {
    return this.questionAnalyses.delete(id);
  }

  // Tool connector methods
  async createToolConnector(connector: InsertToolConnector): Promise<ToolConnector> {
    const newConnector: ToolConnector = {
      id: this.nextId++,
      ...connector,
      createdAt: new Date(),
    };
    this.toolConnectors.set(newConnector.id, newConnector);
    return newConnector;
  }

  async getToolConnectorById(id: number): Promise<ToolConnector | null> {
    return this.toolConnectors.get(id) || null;
  }

  async getToolConnectorsByApplicationId(applicationId: number): Promise<ToolConnector[]> {
    return Array.from(this.toolConnectors.values()).filter(tc => tc.applicationId === applicationId);
  }

  async updateToolConnector(id: number, connector: Partial<InsertToolConnector>): Promise<ToolConnector | null> {
    const existing = this.toolConnectors.get(id);
    if (!existing) return null;
    
    const updated = { ...existing, ...connector };
    this.toolConnectors.set(id, updated);
    return updated;
  }

  async deleteToolConnector(id: number): Promise<boolean> {
    return this.toolConnectors.delete(id);
  }

  // Data collection session methods
  async createDataCollectionSession(session: InsertDataCollectionSession): Promise<DataCollectionSession> {
    const newSession: DataCollectionSession = {
      id: this.nextId++,
      ...session,
      startedAt: null,
      completedAt: null,
    };
    this.dataCollectionSessions.set(newSession.id, newSession);
    return newSession;
  }

  async getDataCollectionSessionById(id: number): Promise<DataCollectionSession | null> {
    return this.dataCollectionSessions.get(id) || null;
  }

  async getDataCollectionSessionsByApplicationId(applicationId: number): Promise<DataCollectionSession[]> {
    return Array.from(this.dataCollectionSessions.values()).filter(dcs => dcs.applicationId === applicationId);
  }

  async updateDataCollectionSession(id: number, session: Partial<InsertDataCollectionSession>): Promise<DataCollectionSession | null> {
    const existing = this.dataCollectionSessions.get(id);
    if (!existing) return null;
    
    const updated = { ...existing, ...session };
    this.dataCollectionSessions.set(id, updated);
    return updated;
  }

  async deleteDataCollectionSession(id: number): Promise<boolean> {
    return this.dataCollectionSessions.delete(id);
  }

  // Audit result methods
  async createAuditResult(result: InsertAuditResult): Promise<AuditResult> {
    const newResult: AuditResult = {
      id: this.nextId++,
      ...result,
      createdAt: new Date(),
    };
    this.auditResults.set(newResult.id, newResult);
    return newResult;
  }

  async getAuditResultById(id: number): Promise<AuditResult | null> {
    return this.auditResults.get(id) || null;
  }

  async getAuditResultsByApplicationId(applicationId: number): Promise<AuditResult[]> {
    return Array.from(this.auditResults.values()).filter(ar => ar.applicationId === applicationId);
  }

  async updateAuditResult(id: number, result: Partial<InsertAuditResult>): Promise<AuditResult | null> {
    const existing = this.auditResults.get(id);
    if (!existing) return null;
    
    const updated = { ...existing, ...result };
    this.auditResults.set(id, updated);
    return updated;
  }

  async deleteAuditResult(id: number): Promise<boolean> {
    return this.auditResults.delete(id);
  }
}

export const storage = new MemStorage();

File: server/storage.py
from sqlalchemy.orm import Session
from sqlalchemy import and_
from typing import List, Optional, Dict, Any
from .database import (
    Application, DataRequest, ToolConnector, DataCollectionSession, 
    QuestionAnalysis, AuditResult, get_db
)
import json

class DatabaseStorage:
    def __init__(self):
        pass
    
    # Applications
    def create_application(self, db: Session, data: dict) -> Application:
        app = Application(**data)
        db.add(app)
        db.commit()
        db.refresh(app)
        return app
    
    def get_application(self, db: Session, app_id: int) -> Optional[Application]:
        return db.query(Application).filter(Application.id == app_id).first()
    
    def get_all_applications(self, db: Session) -> List[Application]:
        return db.query(Application).all()
    
    def update_application(self, db: Session, app_id: int, data: dict) -> Optional[Application]:
        app = db.query(Application).filter(Application.id == app_id).first()
        if app:
            for key, value in data.items():
                setattr(app, key, value)
            db.commit()
            db.refresh(app)
        return app
    
    # Data Requests
    def create_data_request(self, db: Session, data: dict) -> DataRequest:
        request = DataRequest(**data)
        db.add(request)
        db.commit()
        db.refresh(request)
        return request
    
    def get_data_request_by_application_id(self, db: Session, app_id: int) -> Optional[DataRequest]:
        return db.query(DataRequest).filter(DataRequest.applicationId == app_id).first()
    
    def get_data_requests_by_application_id(self, db: Session, app_id: int) -> List[DataRequest]:
        return db.query(DataRequest).filter(DataRequest.applicationId == app_id).all()
    
    # Tool Connectors
    def create_tool_connector(self, db: Session, data: dict) -> ToolConnector:
        connector = ToolConnector(**data)
        db.add(connector)
        db.commit()
        db.refresh(connector)
        return connector
    
    def get_tool_connectors_by_application_id(self, db: Session, app_id: int) -> List[ToolConnector]:
        return db.query(ToolConnector).filter(ToolConnector.applicationId == app_id).all()
    
    def get_tool_connectors_by_ci_id(self, db: Session, ci_id: str) -> List[ToolConnector]:
        return db.query(ToolConnector).filter(ToolConnector.ciId == ci_id).all()
    
    def update_tool_connector_status(self, db: Session, connector_id: int, status: str) -> None:
        connector = db.query(ToolConnector).filter(ToolConnector.id == connector_id).first()
        if connector:
            connector.status = status
            db.commit()
    
    def update_tool_connector(self, db: Session, connector_id: int, data: dict) -> Optional[ToolConnector]:
        connector = db.query(ToolConnector).filter(ToolConnector.id == connector_id).first()
        if connector:
            for key, value in data.items():
                setattr(connector, key, value)
            db.commit()
            db.refresh(connector)
        return connector
    
    def delete_tool_connector(self, db: Session, connector_id: int) -> None:
        connector = db.query(ToolConnector).filter(ToolConnector.id == connector_id).first()
        if connector:
            db.delete(connector)
            db.commit()
    
    # Data Collection Sessions
    def create_data_collection_session(self, db: Session, data: dict) -> DataCollectionSession:
        session = DataCollectionSession(**data)
        db.add(session)
        db.commit()
        db.refresh(session)
        return session
    
    def get_data_collection_session_by_application_id(self, db: Session, app_id: int) -> Optional[DataCollectionSession]:
        return db.query(DataCollectionSession).filter(DataCollectionSession.applicationId == app_id).first()
    
    def update_session_progress(self, db: Session, session_id: int, progress: int, logs: List[Dict]) -> None:
        session = db.query(DataCollectionSession).filter(DataCollectionSession.id == session_id).first()
        if session:
            session.progress = progress
            session.logs = logs
            db.commit()
    
    def update_session_status(self, db: Session, session_id: int, status: str) -> None:
        session = db.query(DataCollectionSession).filter(DataCollectionSession.id == session_id).first()
        if session:
            session.status = status
            db.commit()
    
    # Question Analyses
    def create_question_analysis(self, db: Session, data: dict) -> QuestionAnalysis:
        analysis = QuestionAnalysis(**data)
        db.add(analysis)
        db.commit()
        db.refresh(analysis)
        return analysis
    
    def get_question_analyses_by_application_id(self, db: Session, app_id: int) -> List[QuestionAnalysis]:
        return db.query(QuestionAnalysis).filter(QuestionAnalysis.applicationId == app_id).all()
    
    def update_question_analysis(self, db: Session, analysis_id: int, data: dict) -> Optional[QuestionAnalysis]:
        analysis = db.query(QuestionAnalysis).filter(QuestionAnalysis.id == analysis_id).first()
        if analysis:
            for key, value in data.items():
                setattr(analysis, key, value)
            db.commit()
            db.refresh(analysis)
        return analysis
    
    def delete_question_analyses_by_application_id(self, db: Session, app_id: int) -> None:
        db.query(QuestionAnalysis).filter(QuestionAnalysis.applicationId == app_id).delete()
        db.commit()
    
    # Audit Results
    def create_audit_result(self, db: Session, data: dict) -> AuditResult:
        result = AuditResult(**data)
        db.add(result)
        db.commit()
        db.refresh(result)
        return result
    
    def get_audit_results_by_application_id(self, db: Session, app_id: int) -> List[AuditResult]:
        return db.query(AuditResult).filter(AuditResult.applicationId == app_id).all()
    
    def update_audit_result_status(self, db: Session, result_id: int, status: str, document_path: Optional[str] = None) -> None:
        result = db.query(AuditResult).filter(AuditResult.id == result_id).first()
        if result:
            result.status = status
            if document_path:
                result.documentPath = document_path
            db.commit()

# Global storage instance
storage = DatabaseStorage()

def get_storage():
    return storage